tests/controller_leader_election_test.go

```diff
@@ -0,0 +1,90 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package tests_test
+
+import (
+	"encoding/json"
+	"flag"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	k8sv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/virt-controller/leaderelectionconfig"
+	"kubevirt.io/kubevirt/tests"
+)
+
+var _ = Describe("LeaderElection", func() {
+	flag.Parse()
+
+	virtClient, err := kubecli.GetKubevirtClient()
+	tests.PanicOnError(err)
+
+	BeforeEach(func() {
+		tests.BeforeTestCleanup()
+	})
+
+	Context("Controller pod destroyed", func() {
+		It("should success to start VM", func() {
+			leaderPodName := getLeader()
+			Expect(virtClient.CoreV1().Pods(leaderelectionconfig.DefaultNamespace).Delete(leaderPodName, &metav1.DeleteOptions{})).To(BeNil())
+
+			Eventually(getLeader, 30*time.Second, 5*time.Second).ShouldNot(Equal(leaderPodName))
```

Comment:
I am not yet completely happy with that test. It does not test, if the already waiting pod gets the lease. Since we use a Deployment, it can as well just recover of the respawn of the new pod.

By repeatedly killing every pod which is not the second already running one, we could test that, but maybe you can think of something more elegant.

Reply:
@rmohr Hm, I can get all virt-controller pods before destroying a leader pod, and after destroy I will verify that new leader among old pods, will it be good enough?

Reply:
That sounds good.

Reply:
We might end up with flaky tests then. Maybe we can start two virt-controller pods by hand and kill one. Implementing that might be complex. It would probably involve scaling down the Deployment to 0, taking the pod template and start two pods out of the template ...